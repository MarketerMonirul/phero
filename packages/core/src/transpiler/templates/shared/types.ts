import { JSType, JSValue, SamenManifest } from "../../../domain"

export const promise = (value: JSValue, manifest: SamenManifest) =>
  `Promise<${type(value, manifest)}>`

export const type = (value: JSValue, manifest: SamenManifest): string => {
  return typeToString(value)

  function typeToString(value: JSValue): string {
    switch (value.type) {
      case JSType.number:
      case JSType.string:
      case JSType.boolean:
      case JSType.null:
      case JSType.undefined:
        return value.type

      case JSType.ref:
        return manifest.models[value.id].name

      case JSType.array:
        return `${typeToString(value.elementType)}[]`

      case JSType.tuple:
        return `[${value.elementTypes.map(typeToString).join(", ")}]`

      case JSType.date:
        return "Date"

      case JSType.oneOfTypes:
        return `(${value.oneOfTypes.map(typeToString).join(" | ")})`

      case JSType.object:
        return `{${value.properties
          .map((p) => `${p.name}: ${typeToString(p)}`)
          .join(";")}}`

      case JSType.untyped:
        return "void"
    }
  }
}
