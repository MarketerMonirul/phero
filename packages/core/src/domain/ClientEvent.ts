export enum ClientEvent {
  ClientDidRegister = "CLIENT_DID_REGISTER",
  ManifestDidCHange = "MANIFEST_DID_CHANGE",
}
