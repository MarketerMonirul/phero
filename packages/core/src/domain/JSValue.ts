export type JSValue =
  | JSUntypedValue
  | JSStringValue
  | JSNumberValue
  | JSBooleanValue
  | JSObjectValue
  | JSArrayValue
  | JSDateValue
  | JSNullValue
  | JSUndefinedValue
  | JSOneOfTypesValue
  | JSTupleValue
  | JSModelRef

export enum JSType {
  untyped = "untyped",

  string = "string",
  number = "number",
  boolean = "boolean",

  object = "object",
  array = "array",
  date = "date",

  null = "null",
  undefined = "undefined",

  oneOfTypes = "oneOfTypes",
  tuple = "tuple",

  ref = "ref",
}

export interface JSUntypedValue {
  type: JSType.untyped
}

export interface JSStringValue {
  type: JSType.string
  oneOf?: string[]
}

export interface JSNumberValue {
  type: JSType.number
  oneOf?: number[]
}

export interface JSBooleanValue {
  type: JSType.boolean
  oneOf?: boolean[]
}

export type JSProperty = JSValue & { name: string }

export interface JSObjectValue {
  type: JSType.object
  properties: JSProperty[]
}

export interface JSArrayValue {
  type: JSType.array
  elementType: JSValue
}

export interface JSDateValue {
  type: JSType.date
}

export interface JSNullValue {
  type: JSType.null
}

export interface JSUndefinedValue {
  type: JSType.undefined
}

export interface JSOneOfTypesValue {
  type: JSType.oneOfTypes
  oneOfTypes: JSValue[]
}

export interface JSTupleValue {
  type: JSType.tuple
  elementTypes: JSValue[]
}

export interface JSModelRef {
  type: JSType.ref
  id: string
}
