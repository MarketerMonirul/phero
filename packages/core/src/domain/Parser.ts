export type Parser<T> = (data: unknown) => ParseResult<T>

export type ParseResult<T> = DataParseSuccess<T> | DataParseFailure

export interface DataParseSuccess<T> {
  ok: true
  result: T
}

export interface DataParseFailure {
  ok: false
  errors: DataParseError[]
}

export interface DataParseError {
  message: string
  path?: string
  errors?: DataParseError[]
}
