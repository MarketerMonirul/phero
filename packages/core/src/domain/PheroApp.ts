import ts from "typescript"

export interface PheroApp {
  models: PheroModel[]
  errors: PheroError[]
  services: PheroService[]
}

export interface PheroModel {
  name: string
  ref: Model
}

export type Model =
  | ts.InterfaceDeclaration
  | ts.TypeAliasDeclaration
  | ts.EnumDeclaration

export interface PheroError {
  name: string
  sourceFile: string
  properties: PheroErrorProperty[]
  ref: ts.ClassDeclaration
}

export interface PheroErrorProperty {
  name: string
  type: ts.TypeNode
}

export interface PheroService {
  name: string
  funcs: PheroFunction[]
  config: PheroServiceConfig
  ref: ts.Node
}

export interface PheroFunction {
  name: string
  returnType: ts.TypeNode

  parameters: PheroFunctionParameter[]
  contextParameterType?: ts.TypeNode

  ref: ts.FunctionLikeDeclarationBase
}

export interface PheroFunctionParameter {
  name: string
  questionToken: boolean
  type: ts.TypeNode
}

export interface PheroServiceConfig {
  // these are models which are not used inside functions, but we need them to generate parsers on the server side
  models?: PheroModel[]
  middleware?: PheroMiddlewareConfig[]
  contextType?: ts.TypeNode
}

export interface PheroMiddlewareConfig {
  paramsType: ts.TypeNode
  nextType: ts.TypeNode | undefined
  contextType: ts.TypeNode
  middleware: ts.FunctionLikeDeclarationBase
}
