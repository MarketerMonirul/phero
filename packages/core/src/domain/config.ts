export interface SamenConfig {
  clients: string[]
}

export enum ClientEnvironment {
  Browser,
  Node,
}

export interface ClientConfig {
  development: { url: string }
  production: { url: string }
  env?: ClientEnvironment
}
