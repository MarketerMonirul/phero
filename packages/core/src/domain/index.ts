export * from "./manifest"
export * from "./JSValue"
export * from "./config"
export * from "./ClientEvent"
