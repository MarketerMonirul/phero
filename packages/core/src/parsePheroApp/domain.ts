import ts from "typescript"

export interface PheroApp {
  models: PheroModel[]
  errors: PheroError[]
  services: PheroService[]
}

export interface PheroModel {
  name: string
  ref: Model
}

export type Model =
  | ts.InterfaceDeclaration
  | ts.TypeAliasDeclaration
  | ts.EnumDeclaration

export interface PheroError {
  name: string
  sourceFile: string
  properties: PheroErrorProperty[]
  ref: ts.ClassDeclaration
}

export interface PheroErrorProperty {
  name: string
  type: ts.TypeNode
}

export interface PheroService {
  name: string
  funcs: PheroFunction[]
  config: PheroServiceConfig
  ref: ts.Node
}

export interface PheroFunction {
  name: string
  parameters: ts.ParameterDeclaration[]
  returnType: ts.TypeNode
  serviceContext?: {
    type: ts.TypeNode
    paramName?: string
  }
  ref: ts.FunctionLikeDeclarationBase
}

export interface PheroServiceConfig {
  middleware?: PheroMiddlewareConfig[]
  contextType?: ts.TypeNode
}

export interface PheroMiddlewareConfig {
  paramsType: ts.TypeNode
  nextType: ts.TypeNode | undefined
  contextType: ts.TypeNode
  middleware: ts.FunctionLikeDeclarationBase
}
