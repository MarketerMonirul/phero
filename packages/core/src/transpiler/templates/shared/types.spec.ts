import { JSType } from "../../../domain"
import { type } from "./types"

test("Untyped", () => {
  expect(type({ type: JSType.untyped })).toMatchSnapshot()
})

test("String", () => {
  expect(type({ type: JSType.string })).toMatchSnapshot()
})

test("Number", () => {
  expect(type({ type: JSType.number })).toMatchSnapshot()
})

test("Boolean", () => {
  expect(type({ type: JSType.boolean })).toMatchSnapshot()
})

test("Object", () => {
  expect(
    type({
      type: JSType.object,
      properties: [
        { name: "a", type: JSType.number },
        { name: "b", type: JSType.string },
        { name: "c", type: JSType.ref, id: "Something" },
      ],
    }),
  ).toMatchSnapshot()
})

describe("Array", () => {
  test("Number", () => {
    expect(
      type({ type: JSType.array, elementType: { type: JSType.number } }),
    ).toMatchSnapshot()
  })

  test("Ref", () => {
    expect(
      type({
        type: JSType.array,
        elementType: { type: JSType.ref, id: "Something" },
      }),
    ).toMatchSnapshot()
  })
})

test("Date", () => {
  expect(type({ type: JSType.date })).toMatchSnapshot()
})

test("Null", () => {
  expect(type({ type: JSType.null })).toMatchSnapshot()
})

test("Undefined", () => {
  expect(type({ type: JSType.undefined })).toMatchSnapshot()
})

test("OneOfTypes", () => {
  expect(
    type({
      type: JSType.oneOfTypes,
      oneOfTypes: [
        { type: JSType.number },
        { type: JSType.ref, id: "Something" },
      ],
    }),
  ).toMatchSnapshot()
})

test("Tuple", () => {
  expect(
    type({
      type: JSType.tuple,
      elementTypes: [
        { type: JSType.string },
        { type: JSType.ref, id: "Something" },
      ],
    }),
  ).toMatchSnapshot()
})

test("ModelRef", () => {
  expect(type({ type: JSType.ref, id: "Something" })).toMatchSnapshot()
})
