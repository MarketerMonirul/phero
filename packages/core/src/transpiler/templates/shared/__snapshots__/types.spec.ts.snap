// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array Number 1`] = `"number[]"`;

exports[`Array Ref 1`] = `"Something[]"`;

exports[`Boolean 1`] = `"boolean"`;

exports[`Date 1`] = `"Date"`;

exports[`ModelRef 1`] = `"Something"`;

exports[`Null 1`] = `"null"`;

exports[`Number 1`] = `"number"`;

exports[`Object 1`] = `"{a: number;b: string;c: Something}"`;

exports[`OneOfTypes 1`] = `"(number | Something)"`;

exports[`String 1`] = `"string"`;

exports[`Tuple 1`] = `"[string, Something]"`;

exports[`Undefined 1`] = `"undefined"`;

exports[`Untyped 1`] = `"void"`;
